balancer.go

```diff
@@ -134,31 +134,51 @@ var (
 type Hash struct {
 	rr     RoundRobin
 	Hasher hash.Hash32
+
+	// lock protects Hasher while calculating the hash code.  It is assumed that
+	// the Hasher field is read-only once the Balancer is created, so as a
+	// performance optimization, reads of the field are not protected.
+	lock sync.Mutex
 }
 
-func (h *Hash) Balance(msg Message, partitions ...int) (partition int) {
+func (h *Hash) Balance(msg Message, partitions ...int) int {
 	if msg.Key == nil {
 		return h.rr.Balance(msg, partitions...)
 	}
 
 	hasher := h.Hasher
-	if hasher == nil {
+	shared := hasher != nil
+
+	if shared {
+		h.lock.Lock()
```

Comment:
Alternatively, you could `defer h.lock.Unlock()` here to avoid checking `shared` again below.